File: c/07_auth_failures.c
/* A07 - Auth Failures. VULN: strcmp timing leak + no lockout. */
#include <string.h>

const char *ADMIN_PASS = "admin"; /* VULN: A07 plaintext */

int authenticate(const char *user, const char *pass) {
  if (strcmp(user, "admin") == 0 && strcmp(pass, ADMIN_PASS) == 0)
    return 1;
  return 0;
}

File: python/07_auth_failures.py
"""A07 - Auth Failures. VULN: no lockout, weak session token."""
import random

# VULN: A07 - predictable session ID
def create_session(user_id: int) -> str:
    return f"sess_{user_id}_{random.randint(1000, 9999)}"

def login(db, username, password):
    user = db.find_user(username)
    if user and user.password == password:  # VULN: plaintext compare
        return create_session(user.id)
    return None
